DVCLive: Revisit docs organization and content (#3923)
iterative/dvc.org: content/docs/dvclive/get-started.md

# Get Started

DVCLive is a Python library for logging machine learning metrics and other
metadata in simple file formats which are fully compatible with DVC.

## Set up DVCLive

First of all, you need to add DVCLive to your Python code in order to start

> logging metrics.

Reviewer: I do think this GS approach is better (straight into ML framework integration code sample). But do we still cover generic code application? I guess just in the API ref samples right? (Probably enough)

Reviewer: > do we still cover generic code application? I guess just in the API ref 

I see it's in the [API index page](https://dvc-org-dvclive-refacto-czwjzz.herokuapp.com/doc/dvclive/api-reference) 👍🏼

In this example we are using [Keras](/doc/dvclive/ml-frameworks/keras) for

> training our model and we can add DVCLive as follows:

Reviewer: Note that Hugginfacae and PyTorch seem to be more popular now.

Author: I added tabs with keras, huggingface and pytorch lightning . There is a link right bellow to other frameworks

```python
# train.py
from dvclive.keras import DvcLiveCallback
...
model.fit(
  train_dataset, validation_data=validation_dataset,
  callbacks=[DvcLiveCallback(path="training_metrics")])
```

You can find example snippets and reference of the options available in the
corresponding page of any other of the supported
[ML Frameworks](/doc/dvclive/ml-frameworks).

If you prefer to use the DVCLive API directly, check the
[API Reference](/doc/dvclive/api-reference).

## Set up DVC

> In order to visualize the outputs generated by DVCLive, we can set up a DVC

Reviewer 2: Maybe it's better to say that you need DVC in order to track the outputs and compare experiments. You can already visualize the outputs without DVC.

> pipeline.

Reviewer: Mention this is optional/ recommended?

Author: > Mention this is optional/ recommended?

I think we should treat it as mandatory. Given the current capabilities, I don't think it's realistic to expect people using `dvclive` without a `dvc.yaml`.

This might change as we are discussing new "onboarding" workflows, but for now the above is True.

Reviewer: - I wouldn't undersell DVCLive features (working on its own) but up to you. Maybe just mention it can also work without DVC?
- "set up a DVC pipeline" -> I'd avoid the pipeline concept here (and in general). We're talking about a single stage that's just a required wrapper for `dvc exp`.

Reviewer 2: I agree with your hunch @daavoo that it's unlikely people want to use DVCLive on its own, but I also agree with @jorgeorpinel that we should briefly mention that DVC is optional. I think it's important for clarity. It currently reads like DVCLive will break if you don't do the DVC setup, and it's unclear why.

Author: PTAL to the new version

First, we initialize DVC with `dvc init`:

```dvc
$ dvc init
```

After that, we use `dvc exp init --live` to create a `dvc.yaml` file:

<admon type="info">

Note that the path passed to the `--live` option (`"training_metrics"`) matches
the `path` passed to the callback in the Python code.

</admon>

```dvc
$ dvc exp init \
--live "training_metrics" \
--code "train.py" \
python train.py
```

<details>

### ⚙️ Expand to see the `dvc.yaml`

The `dvc.yaml` will contain a new `train` stage using the DVCLive outputs as
`metrics` and `plots`:

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - train.py
    metrics:
      - training_metrics.json:
          cache: false
    plots:
      - training_metrics/scalars:
          cache: false
```

You can find more details about the structure of files generated by DVCLive in
the [API Reference](/doc/dvclive/api-reference#outputs)

</details>

## Run experiments

You can now

> [run experiments](/doc/user-guide/experiment-management/running-experiments)

Reviewer: Maybe this should be more generic like "Run your code" and mention the non-DVC workflow as well, even if the examples continue with `dvc exp run`.

Reviewer: Also, do we want to cover plain DVC integration (non-experiments i.e. persistent outputs + `dvc repro`) somewhere?

Reviewer 2: I don't think we need to include `dvc repro` since users will get pretty much the same value out of `dvc exp run` here.

with `dvc exp run`:

```dvc
$ dvc exp run
```

## Compare and visualize experiments

After following the above steps we have enabled different ways of monitoring the
training, comparing, and visualizing the experiments:

### DVC

After the experiment has finished, you can use `dvc exp show` / `dvc plots diff`
to compare and visualize experiments.

While the experiment is running, you can monitor the training with the

> [metrics report](/docs/dvclive/api-reference#metrics-report) that DVCLive

Reviewer: Again, I do like this approach a bit more but I feel like it's lacking at least mentioning the vanilla use case (that DVCLive generates a report even if you don't use DVC).

Reviewer 2: Yeah, I'd like to see this at the top before you get to DVC. Show that you can run the code normally and get the live metrics report, then move on to DVC.

Reviewer 2: It has the added benefit of faster time to value: show something useful (the report) almost immediately.

Reviewer: "faster time to value" is the motto of [Get Started type docs](https://www.notion.so/iterative/Higher-level-content-9408f84a9b694fab949feede39142122#8e6381fcf06643aaa1d8df100fd4014a) 🤩

# Get Started

DVCLive is a Python library for logging machine learning parameters, metrics and
other metadata in simple file formats, which is fully compatible with DVC, the
[VS Code Extension](https://marketplace.visualstudio.com/items?itemName=Iterative.dvc),
and [Iterative Studio](https://studio.iterative.ai/).

## Set up DVCLive

First of all, you need to add DVCLive to your Python script:

<toggle>

> <tab title="Keras">

Reviewer: Is there a criteria to the tabs we include here and their order? Just curious...

<img width="693" alt="image" src="https://user-images.githubusercontent.com/1477535/197716872-63187d5a-528e-49d1-a37f-611f21ff07fb.png">

E.g. rn the top 4 with most traffic are: Pytorch, HugginFace, Tensorflow, Keras (in that order).

```python
# train.py
from dvclive.keras import DvcLiveCallback
...
model.fit(
  train_dataset, validation_data=validation_dataset,
  callbacks=[DvcLiveCallback()])
```

</tab>

<tab title="Hugging Face">

```python
# train.py
from dvclive.huggingface import DvcLiveCallback

. . .

 trainer = Trainer(
    model, args,
    train_dataset=train_data,
    eval_dataset=eval_data,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)
trainer.add_callback(DvcLiveCallback())
trainer.train()
```

</tab>
<tab title="Pytorch Lightning">

```python
# train.py
from dvclive.lightning import DvcLiveLogger

. . .
dvclive_logger = DvcLiveLogger()

trainer = Trainer(logger=dvclive_logger)
trainer.fit(model)
```

</tab>

> <tab title="Python API">

Reviewer: Slightly unclear what this means: 1) All the samples are Python; 2) There's no "Python API" under ML Frameworks. Maybe link to https://dvc.org/doc/dvclive/api-reference/live for this one? Currently they all have the same comment under:

> Check the [ML Frameworks](https://dvc.org/doc/dvclive/api-reference/ml-frameworks) page for more details and other supported frameworks.

ways to compare and visualize our experiments.

You can find below an example `dvc.yaml` containing a single stage `train`, that
runs the script from previous steps and tracks the
[DVCLive outputs](/doc/dvclive/outputs) as `dvc metrics` and `dvc plots`:

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - train.py
    metrics:
      - dvclive.json:
          cache: false
    plots:
      - dvclive/scalars:
          cache: false
```

### Run DVC Experiments

You can now
[run experiments](/doc/user-guide/experiment-management/running-experiments)
with `dvc exp run`:

```dvc
$ dvc exp run
Running stage 'train':
> python train.py
...
```

### Compare and Visualize DVC Experiments

After following the above steps, you have enabled different ways of monitoring
the training, comparing, and visualizing the experiments:

#### DVC CLI

You can use `dvc exp show` and `dvc plots` to compare and visualize metrics,
parameters and plots across experiments.

<admon type="info" icon="book">

Learn more in the
[Comparing Experiments](/doc/user-guide/experiment-management/comparing-experiments)
and [Visualizing Plots](/doc/user-guide/experiment-management/visualizing-plots)

> pages of the user guide.

Reviewer: 💅🏼 
```suggestion
pages.
```

</admon>

#### DVC Extension for Visual Studio Code

If you have installed the
[DVC Extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=Iterative.dvc),
you can monitor the training and compare and visualize experiments using the
[`Experiments`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/experiments-table.md)
and
[`Plots`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/plots.md)
views.

![Experiments view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/experiments-table.png)
![Plots view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/plots-trends.png)

> ### Share Results

Reviewer: Seems like this should be an H2 to me (since it's more about Studio than the DVC integration). This would also create an entry in the right side CONTENTS nav.

DVCLive is a Python library for logging machine learning parameters, metrics and

> other metadata in simple file formats, which is fully compatible with DVC.

Reviewer 3: Let's mention VS Code ext here?

Author: Done

## Set up DVCLive

First of all, you need to add DVCLive to your Python code to start logging
metrics:

<toggle>
<tab title="Keras">

```python
# train.py
from dvclive.keras import DvcLiveCallback
...
model.fit(
  train_dataset, validation_data=validation_dataset,
  callbacks=[DvcLiveCallback()])
```

</tab>

<tab title="Hugging Face">

```python
# train.py
from dvclive.huggingface import DvcLiveCallback

. . .

 trainer = Trainer(
    model, args,
    train_dataset=train_data,
    eval_dataset=eval_data,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)
trainer.add_callback(DvcLiveCallback())
trainer.train()
```

</tab>
<tab title="Pytorch Lightning">

```python
# train.py
from dvclive.lightning import DvcLiveLogger

. . .
dvclive_logger = DvcLiveLogger()

trainer = Trainer(logger=dvclive_logger)
trainer.fit(model)
```

</tab>
</toggle>

For other supported frameworks, you can find example snippets and API reference
on the corresponding page of [ML Frameworks](/doc/dvclive/ml-frameworks).

If you prefer to use the DVCLive API directly, check the

> [API Reference](/doc/dvclive/api-reference).

Reviewer 2: Minor: link/mention the basic workflow in the API so it doesn't sound like it's leaving the reader to read through each method.

Reviewer 3: I think we can add a section to the tabs above "No framework" or "Use Directly" ... ? It's worth doing that. I really love these tabs in the get started!

Author: I added a tab named `Python API`, not sure if its a good name

### Run the script

Once you have added DVCLive to our python code, you can run the script as you
would usually do:

```dvc
$ python train.py
```

DVCLive will generate a
[metrics report](/doc/dvclive/api-reference/live/make_report) in
`dvclive/report.html` containing all the logged data. It will be automatically
updated during training on each `step` update:

![HTML report](/img/dvclive-html.gif).

You can find more details about the structure of files generated by DVCLive in
the [Output folder structure](/doc/dvclive/outputs).

## Set up DVC

To build on the previous section, you can optionally wrap our code with DVC to
run [DVC experiments](/doc/user-guide/experiment-management/) and enable new
ways to compare and visualize our experiments.

> 

Reviewer 2: Hopefully, we can improve this section soon. For now, what do you about making the entire setup an expandable section? Realistically, right now a new user probably needs to go the DVC docs and get set up with DVC before coming back to DVCLive, so I don't think it's worth pushing down the other content to explain how to setup DVC.

Author: I still want to highlight the `dvc.yaml` syntax required (`metrics`, `plots`) to get the things rendered. 
Looking at it right now I don't really like that the `dvc.yaml` is hidden in an expandable section.

Author: I removed `dvc init` and `exp init` but still displayed the `dvc.yaml`. Still think it makes sense as minimum for building the sections below about different visualization

First, you can initialize DVC with `dvc init`:

```dvc
$ dvc init
```

After that, you can use `dvc exp init --live` to create a `dvc.yaml` file:

<admon type="info">

Note that the path passed to the `--live` option (`"dvclive"`) matches the
`path` passed to [`Live()`](/doc/dvclive/api-reference/live) in the Python code
(`"dvclive"` is used as default when no value is passed).

</admon>

```dvc

> $ dvc exp init \

Reviewer 2: ```suggestion
$ dvc exp init --explicit \
```

Otherwise I get additional stuff in dvc.yaml compared to what you show below.

    cmd: python train.py
    deps:
      - train.py
    metrics:
      - dvclive.json:
          cache: false
    plots:
      - dvclive/scalars:
          cache: false
```

If you log other types of data you can check the
[Output folder structure](/doc/dvclive/outputs) and update your `dvc.yaml`
accordingly.

</details>

## Run experiments

You can now
[run experiments](/doc/user-guide/experiment-management/running-experiments)
with `dvc exp run`:

```dvc
$ dvc exp run
```

## Compare and visualize experiments

After following the above steps, you have enabled different ways of monitoring
the training, comparing, and visualizing the experiments:

### DVC

The [metrics history](/doc/dvclive/api-reference/live/log#step-updates)
`dvclive/scalars` can be visualized with `dvc plots`.

The [metrics summary](/doc/dvclive/api-reference/live/log#description) in
`dvclive.json` can be visualized with `dvc metrics`.

If you have logged parameters with `Live.log_param()`, they will be stored in
`dvclive/params.yaml` and can be visualized with `dvc params`.

In addition, the `dvc exp show` and `dvc exp diff` commands can be used to
compare both metrics and parameters across experiments.

> 

Reviewer 2: I don't think it's worth going through every related DVC command here. Mentioning `dvc exp show` and `dvc plots show/diff` is probably enough.

### Visual Studio Code

If you have installed the
[DVC Extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=Iterative.dvc),
you can monitor the training and compare and visualize experiments using the
[`Experiments`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/experiments-table.md)
and
[`Plots`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/plots.md)
views.

![Experiments view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/experiments-table.png)
![Plost view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/plots-trends.png)

### Iterative Studio

After the experiment has finished and you have committed and pushed the results,
[Iterative Studio](/doc/studio) will automatically parse the outputs generated
by DVCLive, allowing you to

> [compare and visualize experiments](/doc/studio/user-guide/projects-and-experiments/visualize-and-compare):

Reviewer 2: To distinguish from the other products above, it might be better to focus on sharing experiments instead of comparing and visualizing.

    live.next_step()
```

</tab>
</toggle>

Check the [ML Frameworks](/doc/dvclive/api-reference/ml-frameworks) page for
more details and other supported frameworks.

If you prefer to use DVCLive directly, check the overview of the

> [Python API](/doc/dvclive/api-reference).

Reviewer 2: Minor/not blocking, but you could drop this now that you added the Python API (which looks good IMO).

### Run the script

Once you have added DVCLive to our python code, you can run the script as you
would usually do:

```dvc
$ python train.py
```

DVCLive will generate a [report](/doc/dvclive/outputs#report) in
`dvclive/report.html` containing all the logged data. It will be automatically
updated during training on each `step` update:

![HTML report](/img/dvclive-html.gif).

You can find more details about the structure of files generated by DVCLive in

> the [Output folder structure](/doc/dvclive/outputs).

Reviewer 2: It might be worth showing some example output here, like `dvclive.json` and `dvclive/scalars`, especially since you mention them below without any prior context on this page.

## Set up DVC

To build on the previous section, you can optionally wrap our code with DVC to
run [DVC experiments](/doc/user-guide/experiment-management/) and enable new
ways to compare and visualize our experiments.

> You can find below how an example `dvc.yaml` containing a single sage `train`,

Reviewer 2: ```suggestion
You can find below an example `dvc.yaml` containing a single stage `train`,
```

that runs the script from previous steps and tracks the
[DVCLive outputs](/doc/dvclive/outputs) as `dvc metrics` and `dvc plots`:

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - train.py
    metrics:
      - dvclive.json:
          cache: false
    plots:
      - dvclive/scalars:
          cache: false
```

## Run experiments

You can now
[run experiments](/doc/user-guide/experiment-management/running-experiments)
with `dvc exp run`:

```dvc
$ dvc exp run
Running stage 'train':
> python train.py
...
```

> ## Compare and visualize experiments

Reviewer 2: ```suggestion
## Compare and visualize DVC experiments
```

Somewhere in the header or text below, I think it's worth clarifying that comparing experiments is part of DVC integration and not standalone DVCLive functionality.

After following the above steps, you have enabled different ways of monitoring
the training, comparing, and visualizing the experiments:

> ### DVC

Reviewer 2: Should this be `CLI` or `DVC CLI`? The VS Code extension is also DVC.

> You can use `dvc exp show` and `dvc plots` to compare and visualize metrics,

Reviewer 2: Link to https://dvc.org/doc/user-guide/experiment-management/comparing-experiments and/or https://dvc.org/doc/user-guide/experiment-management/visualizing-plots?

parameters and plots across experiments.

### Visual Studio Code

If you have installed the
[DVC Extension for Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=Iterative.dvc),
you can monitor the training and compare and visualize experiments using the
[`Experiments`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/experiments-table.md)
and
[`Plots`](https://github.com/iterative/vscode-dvc/blob/main/extension/resources/walkthrough/plots.md)
views.

![Experiments view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/experiments-table.png)

> ![Plost view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/plots-trends.png)

Reviewer 2: ```suggestion
![Plots view](https://github.com/iterative/vscode-dvc/raw/main/extension/resources/walkthrough/images/plots-trends.png)
```
